src/ServiceControl/Monitoring/HeartbeatMonitor.cs

```diff
@@ -0,0 +1,53 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using System.Threading;
+
+    class HeartbeatMonitor
+    {
+        volatile RecordedHeartbeat heartbeat = new RecordedHeartbeat(HeartbeatStatus.Unknown, null);
+
+        public void MarkAlive(DateTime timestamp)
+        {
+            var newReading = new RecordedHeartbeat(HeartbeatStatus.Alive, timestamp);
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+
+                if (priorReading.Timestamp.GetValueOrDefault() <= timestamp)
+                {
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
+                }
+                else
+                {
+                    done = true;
+                }
+            }
+        }
+
+        public RecordedHeartbeat Age(DateTime cutoff)
+        {
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+                if (priorReading.Timestamp.GetValueOrDefault() < cutoff)
+                {
+                    var newReading = new RecordedHeartbeat(HeartbeatStatus.Dead, priorReading.Timestamp);
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
```

Comment:
Ah, ok. Now it makes sense. I see it is mainly for concurrent marking as alive. At first I only thought about the race conditions between marking alive and dead. I ❤️ it.